### PR TITLE
Feature/modify list all items

### PR DESCRIPTION
**Pull request by dieDeMiguel**
Merging changes in index.js, .gitignore and items.csv

### CODE REVIEW

File: routes/index.js [hunk 1]
```diff
@@ -10,22 +12,58 @@ router.get('/', function(req, res, next) {
...
 
 //Create an Item
 router.post('/items', function(req, res, next) {
-  let item = JSON.stringify({'id': req.body.id, 'name': req.body.name});
-  
-  fs.appendFileSync('items.txt', item, (err) => {
-    if(err) throw err;
-  console.log("Item was saved into memory");
-  });
+  if(req.body.id && req.body.name) {
```
Comment (charlitos85) [inline routes/index.js:29]: Move this validations to a function

File: routes/index.js [hunk 2]
```diff
@@ -10,22 +12,58 @@ router.get('/', function(req, res, next) {
 
 // GET objects array complete
 router.get('/items', function(req, res, next) {
-  res.send('Here you will find the array parsed');
+    const readFile = 'items.csv'
+      fs.readFile(readFile, (err, file) => {
+        if (err) {
+          const error = new Error('The file doesn\'t exist');
+          error.status = 417,
+          next(error);
```
Comment (charlitos85) [inline routes/index.js:20]: This is not the right way

File: routes/index.js [hunk 3]
```diff
@@ -1,6 +1,8 @@
 var express = require('express');
+var appToUse = express()
```
Comment (charlitos85) [inline routes/index.js:2]: Const router

File: routes/index.js (hunk 2 above)
Comment (dieDeMiguel) [reply routes/index.js:20]: throw new Error("Here goes a message")

Something like this?

File: routes/index.js (hunk 1 above)
Comment (dieDeMiguel) [reply routes/index.js]: done, I call it idNameValidation and it's saved under the controllers folder

File: routes/index.js (hunk 2 above)
Comment (dieDeMiguel) [reply routes/index.js]: Error handling now uses callbacks

File: routes/index.js (hunk 3 above)
Comment (dieDeMiguel) [reply routes/index.js]: done

File: controllers/readfile.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+                   name: data.name
+            }
+           })
+       }
+    })
+}
+  
+
+
+module.exports = readFile
```
Comment (charlitos85) [inline controllers/readfile.js:21]: missing `;`

File: controllers/readfile.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+const fs = require('fs');
+const { call } = require('../app');
```
Comment (charlitos85) [inline controllers/readfile.js:2]: this is not required

File: controllers/storeitem.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+const fs = require('fs');
+
+const storeItem = (id, name, callback) => {
```
Comment (charlitos85) [inline controllers/storeitem.js:3]: should to be an object be more abstract

```
function saveItem( item, callback) {
...
}
```

File: controllers/storeitem.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    newItem = JSON.stringify({'id': id, 'name': name});
```
Comment (charlitos85) [inline controllers/storeitem.js:4]: missing the definition `const newItem`

File: controllers/storeitem.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    const fileName = 'items.csv'
+    fs.writeFile(fileName, newItem, (error) => {
```
Comment (charlitos85) [inline controllers/storeitem.js:6]: missing endcoding

File: controllers/storeitem.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+      if(error) {
+        callback('Error', error);
```
Comment (charlitos85) [inline controllers/storeitem.js:8]: this is wrong should to be
```return callback(error);```

File: controllers/validation.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+const fs = require('fs');
+const storeItem = require('./storeitem');
+  
+  const idNameValidation = (id, name, callback) => {
```
Comment (charlitos85) [inline controllers/validation.js:4]: wrong tabulation...the function should to received a obj. the name of the function not make sense should to be something like `validateItem`

File: controllers/validation.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    if(id && name.length > 3) {
+      storeItem(id, name, (error, response) => {
+        if(error) {
+          callback(error, undefined);
```
Comment (charlitos85) [inline controllers/validation.js:8]: this is wrong

File: routes/index.js [hunk 12]
```diff
@@ -1,31 +1,42 @@
 var express = require('express');
 var router = express.Router();
 const fs = require('fs');
+const app = require('../app');
```
Comment (charlitos85) [inline routes/index.js:4]: this is not required.

File: routes/index.js [hunk 13]
```diff
@@ -1,31 +1,42 @@
...
+const idNameValidation = require('../controllers/validation');
```
Comment (charlitos85) [inline routes/index.js:5]: this is not at controller. its an util or helper

File: routes/index.js [hunk 14]
```diff
@@ -1,31 +1,42 @@
...
-  fs.appendFileSync('items.txt', item, (err) => {
-    if(err) throw err;
-  console.log("Item was saved into memory");
-  });
-});
+//Create an Item with error handling
+router.post('/items', function(req, res) {
+  idNameValidation(req.body.id, req.body.name, (error, response) => {
+    if(error) {
+      res.send({
```
Comment (charlitos85) [inline routes/index.js:30]: res.status(500).send(response)

File: controllers/storeitem.js (hunk 6 above)
Comment (dieDeMiguel) [reply controllers/storeitem.js:3]: done

File: controllers/storeitem.js (hunk 7 above)
Comment (dieDeMiguel) [reply controllers/storeitem.js:4]: done

File: controllers/validation.js (hunk 10 above)
Comment (dieDeMiguel) [reply controllers/validation.js:4]: done

File: controllers/validation.js (hunk 11 above)
Comment (dieDeMiguel) [reply controllers/validation.js:8]: moved item info validation  to the storItem function

File: routes/index.js (hunk 14 above)
Comment (dieDeMiguel) [reply routes/index.js:29]: or maybe res.status(500).send(error)

File: routes/index.js (hunk 13 above)
Comment (dieDeMiguel) [reply routes/index.js]: changed folder name

File: routes/index.js (hunk 12 above)
Comment (dieDeMiguel) [reply routes/index.js]: removed

File: utils/validation.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
+const fs = require('fs');
+const storeItem = require('./storeitem');
+  
+const validateItem = (item, callback) => {
```
Comment (charlitos85) [inline utils/validation.js:4]: this function only need to validate and return the erros

File: utils/storeitem.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
+const fs = require('fs');
+
+function saveItem(item, callback) {
```
Comment (charlitos85) [inline utils/storeitem.js:3]: this function only need to save a item, shoudn't validate nothing

File: utils/readfile.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
+const fs = require('fs');
+
+const readFile = (fileName, callback) => {
```
Comment (charlitos85) [inline utils/readfile.js:3]: should to be a regular funcion

File: utils/readfile.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+           callback(null, {
+               item : {
+                   id: data.id, 
+                   name: data.name
+            }
+           })
+       }
+    })
+}
+
```
Comment (charlitos85) [inline utils/readfile.js:17]: extra spaces

File: utils/readfile.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
+const fs = require('fs');
+
+const readFile = (fileName, callback) => {
+   fs.readFile(fileName, (error, data) => {
+       if(error) {
+           callback('The file couldn\'t be read', null);
+       } else {
+           callback(null, {
```
Comment (charlitos85) [inline utils/readfile.js:8]: callback(null, data)

File: utils/validation.js (hunk 15 above)
Comment (charlitos85) [reply utils/validation.js:4]: and this function should to be sync

File: routes/index.js [hunk 20]
```diff
@@ -1,31 +1,41 @@
...
   res.render('index', { title: 'Express' });
 });
 
 
-// GET objects array complete
-router.get('/items', function(req, res, next) {
-  res.send('Here you will find the array parsed');
+// GET items 
+router.get('/items', function(req, res) {
+    const fileName = 'items.csv'
```
Comment (charlitos85) [inline routes/index.js:16]: move all this to a controller, its not responsability of the router to handle logic

File: controllers/validation.js [hunk 21]
```diff
@@ -0,0 +1,15 @@
+const fs = require('fs');
+const storeItem = require('./storeitem');
+  
+const validateItem = (item, callback) => {
+    storeItem(item, (error, response) => {
+      if(error) {
+        callback(error, null);
```
Comment (charlitos85) [inline controllers/validation.js:7]: return callback(error); don't need the null at the end

File: controllers/storeitem.js [hunk 22]
```diff
@@ -0,0 +1,30 @@
+const fs = require('fs');
+
+function saveItem(item, callback) {
+    const itemName = item.name.toLowerCase().trim();
```
Comment (charlitos85) [inline controllers/storeitem.js:4]: used the validation function if you want to check, but. its not the responsability of this. function do that, only need to save a item into the file

File: controllers/readfile.js [hunk 23]
```diff
@@ -0,0 +1,20 @@
+const fs = require('fs');
+
+const readFile = (fileName, callback) => {
```
Comment (charlitos85) [inline controllers/readfile.js:3]: regular function

File: controllers/readfile.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+                   name: data.name
+            }
+           })
+       }
+    })
+}
+  
+
+
+module.exports = readFile;
```
Comment (charlitos85) [inline controllers/readfile.js:20]: missing ;

File: controllers/readfile.js [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+           callback(null, {
+               item : {
+                   id: data.id, 
+                   name: data.name
+            }
+           })
+       }
+    })
+}
+
```
Comment (charlitos85) [inline controllers/readfile.js:17]: extra spaces

File: controllers/validation.js (hunk 21 above)
Comment (dieDeMiguel) [reply controllers/validation.js]: done

File: controllers/storeitem.js (hunk 22 above)
Comment (dieDeMiguel) [reply controllers/storeitem.js]: changed

File: controllers/readfile.js (hunk 23 above)
Comment (dieDeMiguel) [reply controllers/readfile.js]: done

File: controllers/readfile.js (hunk 24 above)
Comment (dieDeMiguel) [reply controllers/readfile.js]: added

File: controllers/readfile.js (hunk 25 above)
Comment (dieDeMiguel) [reply controllers/readfile.js]: reviewed

File: utils/readfile.js (hunk 18 above)
Comment (dieDeMiguel) [reply utils/readfile.js:12]: removed

File: utils/readfile.js (hunk 19 above)
Comment (dieDeMiguel) [reply utils/readfile.js]: done

**Pull request merged**